### PR TITLE
[Hotfix] Dashboard authentication on Py3

### PR DESCRIPTION
**Pull request by Anonym-tsk**
## Description:
Fixed dashboard authentication with Py3
Link: #668 

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by Anonym-tsk** [comment]
Please merge this fix to 1.14 too

File: esphome/dashboard/dashboard.py [hunk 1]
```diff
@@ -57,6 +57,9 @@ def parse_args(self, args):
         if not self.on_hassio:
             self.username = args.username or os.getenv('USERNAME', '')
             self.using_password = bool(password)
+            if not IS_PY2:
+                self.username = str(self.username.encode())
```
Comment (OttoWinter) [inline esphome/dashboard/dashboard.py]: This does not look right. You're converting a `str` to `bytes` (via the `.encode()`). And then converting the `bytes` object back to `str`, which will leave `b''` artifacts. Have you tested this hotfix?

```bash
$ python3
>> user = 'hello'
>>> str(user.encode())
"b'hello'"
```

**Review by Anonym-tsk — COMMENTED** [review]
Yes, i tested this code on py2 and py3.
I don't know why, but here the same code https://github.com/Anonym-tsk/esphome/blob/ab3a5669153cd2a53787ce570b2204e594c0a5d5/esphome/dashboard/dashboard.py#L634

**Comment by OttoWinter** [comment]
Yeah that's where the real error is. That conversion there is just wrong.

The following patch will work (and clean up the unicode handling in this file):

```diff
diff --git a/esphome/dashboard/dashboard.py b/esphome/dashboard/dashboard.py
index 0927fe2d..c934626d 100644
--- a/esphome/dashboard/dashboard.py
+++ b/esphome/dashboard/dashboard.py
@@ -29,7 +29,7 @@ import tornado.websocket
 from esphome import const, util
 from esphome.__main__ import get_serial_ports
 from esphome.helpers import mkdir_p, get_bool_env, run_system_command
-from esphome.py_compat import IS_PY2, decode_text
+from esphome.py_compat import IS_PY2, decode_text, encode_text
 from esphome.storage_json import EsphomeStorageJSON, StorageJSON, \
     esphome_storage_path, ext_storage_path, trash_storage_path
 from esphome.util import shlex_quote
@@ -85,12 +85,11 @@ class DashboardSettings(object):
     def check_password(self, username, password):
         if not self.using_auth:
             return True
+        if username != self.username:
+            return False
 
-        if IS_PY2:
-            password = hmac.new(password).digest()
-        else:
-            password = hmac.new(password.encode()).digest()
-        return username == self.username and hmac.compare_digest(self.password_digest, password)
+        password_digest = hmac.new(encode_text(password)).digest()
+        return hmac.compare_digest(self.password_digest, password_digest)
 
     def rel_path(self, *args):
         return os.path.join(self.config_dir, *args)
@@ -610,8 +609,8 @@ class LoginHandler(BaseHandler):
             'X-HASSIO-KEY': os.getenv('HASSIO_TOKEN'),
         }
         data = {
-            'username': str(self.get_argument('username', '')),
-            'password': str(self.get_argument('password', ''))
+            'username': decode_text(self.get_argument('username', '')),
+            'password': decode_text(self.get_argument('password', ''))
         }
         try:
             req = requests.post('http://hassio/auth', headers=headers, data=data)
@@ -628,8 +627,8 @@ class LoginHandler(BaseHandler):
         self.render_login_page(error="Invalid username or password")
 
     def post_native_login(self):
-        username = str(self.get_argument("username", '').encode('utf-8'))
-        password = str(self.get_argument("password", '').encode('utf-8'))
+        username = decode_text(self.get_argument("username", ''))
+        password = decode_text(self.get_argument("password", ''))
         if settings.check_password(username, password):
             self.set_secure_cookie("authenticated", cookie_authenticated_yes)
             self.redirect("/")
```

**Comment by Anonym-tsk** [comment]
Thank you for the patch. I tested and it works on Py2 and Py3.

**Pull request merged**